hudi-common/src/main/java/org/apache/hudi/common/model/HoodieAvroRecord.java

```diff
@@ -189,14 +193,37 @@ public Option<Map<String, String>> getMetadata() {
 
   @Override
   public Option<HoodieAvroIndexedRecord> toIndexedRecord(Schema recordSchema, Properties props) throws IOException {
-    Option<IndexedRecord> avroData = getData().getInsertValue(recordSchema, props);
+    Option<IndexedRecord> avroData = getCachedDeserializedRecord(recordSchema, props);
     if (avroData.isPresent()) {
       return Option.of(new HoodieAvroIndexedRecord(avroData.get()));
     } else {
       return Option.empty();
     }
   }
 
+  private Option<IndexedRecord> getCachedDeserializedRecord(Schema recordSchema, Properties props) throws IOException {
+    // Check schema identical
+    if (this.cachedDeserializedRecord != null && this.cachedDeserializedRecord.isPresent()
+        && !compareSchema(cachedDeserializedRecord.get().getSchema(), recordSchema)) {
+      this.cachedDeserializedRecord = null;
+    }
+    if (this.cachedDeserializedRecord == null) {
+      this.cachedDeserializedRecord = this.data.getInsertValue(recordSchema, props);
+    }
+    return this.cachedDeserializedRecord;
+  }
+
+  private static Boolean compareSchema(Schema left, Schema right) {
+    if (left == null || right == null) {
+      return false;
+    }
+    Pair<Schema, Schema> schemaPair = Pair.of(left, right);
+    if (!SCHEMA_COMPARE_MAP.containsKey(schemaPair)) {
```

Comment:
I don't think this setup makes sense -- complexity of comparing the `Pair(leftSchema, rightSchema)` is def more than just comparing 2 schemas.

Instead, let's just keep `Pair<Schema, GenericRecord>` cached and compare it whenever we retrieve

Reply:
1. SCHEMA_COMPARE_MAP is hashmap and its get/containsKey function is O(1). We just compare Pair once.
2. As you said before, schema compare will kill perf-gain. We need to cache the compare result.

Reply:
You're right it's O(1), but i'm not talking about asymptotic complexity i'm talking about just the cost of looking up in the HashMap: you use `Pair<Schema, Schema>` as a key and that means that we will have to do the equality check on the HM entry when we look it up, and in that case it would have to compare 2 schemas to retrieve the previous result of their comparison which obviously doesn't make sense.